File: examples/matmul/matmul.cpp
// g++ block_matrix_mul.cpp -O3 -o block_matrix_mul
#include <bits/stdc++.h>
#include <sys/time.h>

void Multiply(int n, double **a, double **b, double **c) {
    int blockSize = floor(sqrt(n));

    for (int bi = 0; bi < n; bi += blockSize)
        for (int bj = 0; bj < n; bj += blockSize)
            for (int bk = 0; bk < n; bk += blockSize)
                for (int i = 0; i < blockSize; i++)
                    for (int j = 0; j < blockSize; j++)
                        for (int k = 0; k < blockSize; k++)
                            c[bi + i][bj + j] += a[bi + i][bk + k] * b[bk + k][bj + j];
}

int main(int argc, char *argv[]) {
    int n = 32 * 32;
    int numreps = 1; // 1

    if (argc >= 2) {
        n = atoi(argv[1]);
    }
    if (argc >= 3) {
        numreps = atoi(argv[2]);
    }

    ///////////////////// Matrix A //////////////////////////

    double **A = (double **)malloc(n * sizeof(double *));
    A[0] = (double *)malloc(n * n * sizeof(double));
    if (!A) {
        printf("memory failed \n");
        exit(1);
    }
    for (int i = 1; i < n; i++) {
        A[i] = A[0] + i * n;
        if (!A[i]) {
            printf("memory failed \n");
            exit(1);
        }
    }

    ///////////////////// Matrix B //////////////////////////

    double **B = (double **)malloc(n * sizeof(double *));
    B[0] = (double *)malloc(n * n * sizeof(double));
    if (!B) {
        printf("memory failed \n");
        exit(1);
    }
    for (int i = 1; i < n; i++) {
        B[i] = B[0] + i * n;
        if (!B[i]) {
            printf("memory failed \n");
            exit(1);
        }
    }

    ///////////////////// Matrix C //////////////////////////
    double **C = (double **)malloc(n * sizeof(double *));
    C[0] = (double *)malloc(n * n * sizeof(double));
    if (!C) {
        printf("memory failed \n");
        exit(1);
    }
    for (int i = 1; i < n; i++) {
        C[i] = C[0] + i * n;
        if (!C[i]) {
            printf("memory failed \n");
            exit(1);
        }
    }

    // initialize the matrices
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            A[i][j] = 1.0 * rand() / RAND_MAX;
            B[i][j] = 1.0 * rand() / RAND_MAX;
        }
    }

    // multiply matrices
    struct timeval tv1, tv2;
    double elapsed;
    printf("Multiply matrices %d times...\n", numreps);
    for (int i = 0; i < numreps; i++) {
        gettimeofday(&tv1, NULL);
        Multiply(n, A, B, C);
        gettimeofday(&tv2, NULL);
        double delay =
            (double)(tv2.tv_sec - tv1.tv_sec) + (double)(tv2.tv_usec - tv1.tv_usec) * 1.e-6;
        printf("time = %lf\n", delay);
        elapsed += (double)(tv2.tv_sec - tv1.tv_sec) + (double)(tv2.tv_usec - tv1.tv_usec) * 1.e-6;
    }
    printf("Total time = %lf, %d runs, avg time = %lf\n", elapsed, numreps, elapsed / numreps);

    // deallocate memory

    free(A[0]);
    free(A);
    free(B[0]);
    free(B);
    free(C[0]);
    free(C);
    return 0;
}

File: examples/matmul/matmul.py
import math
import time
import sys
import random
# import numba
# from numba.typed import List

# @numba.jit(nopython=True)
def matrix_mul(n, matrix_A, matrix_B, matrix_C):
    block_size = int(math.floor(math.sqrt(n)))
    for bi in range(0, n, block_size):
        for bj in range(0, n, block_size):
            for bk in range(0, n, block_size):
                for i in range(block_size):
                    for j in range(block_size):
                        for k in range(block_size):
                            matrix_C[bi + i][bj + j] += matrix_A[bi + i][bk + k] * matrix_B[bk + k][bj + j]

def block_matrix_mul(n, numreps):
    # matrix_A = List([List([random.random() for i in range(n)]) for j in range(n)])
    # matrix_B = List([List([random.random() for i in range(n)]) for j in range(n)])
    # matrix_C = List([List([0.0 for i in range(n)]) for j in range(n)])
    matrix_A = [[random.random() for i in range(n)] for j in range(n)]
    matrix_B = [[random.random() for i in range(n)] for j in range(n)]
    matrix_C = [[0.0 for i in range(n)] for j in range(n)]

    print('Multiply calculation %d times' % numreps)
    total_elapsed = 0
    for epoch in range(numreps):
        start = time.time()
        matrix_mul(n, matrix_A, matrix_B, matrix_C)
        elapsed = time.time() - start
        print("Matmul Time = %f, epoch = %d/%d" % (elapsed, epoch, numreps))
        total_elapsed += elapsed
    print("Total time = %f, %d runs, avg time = %f\n" %
          (total_elapsed, numreps, total_elapsed / numreps));

if __name__ == "__main__":
    n = 32 * 32
    numreps = 1

    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    if len(sys.argv) > 2:
        numreps = int(sys.argv[2])

    block_matrix_mul(n, numreps)
